### PR TITLE
Feature/refactor generate

### PR DESCRIPTION
**Pull request by michalk8**
This PR refactors some choices in the `generate.py` file and fixes (in some part) the issues below:

- #1689: can be closed.
insert_segments validates option lengths up front and commits only after every sequence is built; stream_generate rejects max_tokens=0 rather than crashing. Raising is the intended contract, not the empty stream the issue suggests.
- #1681: batch half addressed, stream half rejected.
Ours names the offending option and is atomic. Its max_tokens < 0 guard would close off the working unbounded range (verified at -1 and -7).
- #1708: overridden by design.
Same crash fixed, opposite contract: max_tokens=0 raises, because it already means prefill-only in generate_step and an empty stream would silently skip the prefill.
- #1668: addressed.
On more surface too — BatchGenerator.insert, not just batch_generate.
- #1657 / issue #1472: Closed.
_normalize_logits_processors turns None into () at insert, so it can't reach the per-token loop.
- #1645: addressed.
Guarded tps properties in generate_utils.py; stats()'s finally only accumulates, so it can't mask the body's exception.
- #1631: partially addressed.
keep > 0 guard moved into _merge_caches; the unconditional keep=4 blocking --kv-bits stands.

Also fixes some issues in the KV cache's left padding (syncing too often); now does ` self.left_padding.tolist()[idx]` instead of computing this using `mx`).

### CODE REVIEW

File: mlx_lm/models/cache.py [hunk 1]
```diff
@@ -1023,8 +1023,8 @@ def filter(self, batch_indices):
         self.offset = self.offset[batch_indices]
         self.left_padding = self.left_padding[batch_indices]
 
-        # Shift left to reduce padding
-        min_left_pad = self.left_padding.min().item()
+        mx.eval(self.left_padding)
+        min_left_pad = min(self.left_padding.tolist())
```
Comment (zcbenz) [inline mlx_lm/models/cache.py:1027]: What this change does is moving computation of min from gpu to cpu?

**Review by zcbenz — COMMENTED** [review]
Is it possible to split the PR into smaller ones?

**Comment by michalk8** [comment]
> Is it possible to split the PR into smaller ones?

Yes, splitting this!

File: mlx_lm/models/cache.py (hunk 1 above)
Comment (michalk8) [reply mlx_lm/models/cache.py:1027]: I explain it here #1824

**Comment by michalk8** [comment]
Closing this in favor of the linked PRs.
